### ISSUE TITLE
🐛 Bug Report: Secrets are visible in browser address bar after pressing "Start Over" in Scaffolder wizard

### ISSUE THREAD

**Issue by jomeier** [backstage/backstage#24483]
### 📜 Description

Hi,
in Version 1.25.2 I created a scaffolder template, where I use a password.
After pressing "create" and waiting for the steps to end, I pressed "Start Over". Afterwards you can see my form values in the browser address bar. Passwords can easily be seen as URL parameters (see screenshot).
This could be a problem, if - for example - you make a screen share or record a training video.
Some ideas of how to fix that?
Thanks and greetings,
Josef

### 👍 Expected behavior

Passwords shouldn't be visible in the Browsers address bar.

### 👎 Actual Behavior with Screenshots

![image](https://github.com/backstage/backstage/assets/12024238/ab10f16c-4b11-4a92-89e1-a0f4e839e7f3)

### 👟 Reproduction steps

Use a Secret field in Scaffolder template, fill out everything, Create and "Start Over". You will see all parameters, also the secret and its value, in the browser address bar.

Template snippet of the secret parameter:
```
        password:
          title: Password
          default: Password
          type: string
          description: Password of Service Account.
          ui:autofocus: true
          ui:field: password
          ui:widget: password
```

### 🖥️ Your Environment

Recent Chrome Browser

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by benjdlambert** [comment]
Hey @jomeier! Thanks for raising this!

I think if you're wanting to collect secrets from the frontend, you're going to be best off following this guide: https://backstage.io/docs/features/software-templates/writing-templates/#using-secrets

TLDR is you'll want to use `ui:field: Secret` in order to protect these secrets. We should probably look at overriding the default `ui:widget: password` and `ui:field: password` so that it's clear that these are not secure.

**Comment by benjdlambert** [comment]
I wonder if we can make this clearer for users that the `ui:widget: password` is not secure. I'm thinking that there's two options:

- We replace the `ui:widget: password` with the `ui:field: Secret` and it just works, or we render red text saying that this is not secure and you should use `ui:field: Secret` instead.
- We reject templates in the processor that have `ui:widget: password`

**Comment by Rugvip** [comment]
We also discussed to option of simply overriding the password widget with a plain text one. Potentially with a warning somewhere to avoid confusion

**Comment by jomeier** [comment]
One concern I have with that is, that the secret must be accessible in the form context to use it in custom ui field extensions:
https://backstage.io/docs/features/software-templates/writing-custom-field-extensions/#access-data-from-other-fields
With ui:widget: Secret passwords are secure, but can't be accessed in the form. I only get '****' characters instead of the secret in my code and not the secret value itself. Or did I miss the right spot, where I can get the value of fields marked with "ui:field: secret" ?

**Comment by benjdlambert** [comment]
@jomeier no, with `ui:widget` passwords are unfortunately not treated as secure, they are not redacted in logs, and could be accessed / logged through API's. When you use `ui:field: Secret` you will need to use `${{ secrets.PARAM_NAME }}` not `${{ parameters.PARAM_NAME }}` instead.

**Comment by benjdlambert** [comment]
I'm gonna whip up a PR to replace the `ui:widget: password` with our our own `secret` implementation, but I have some concerns that we're going to break existing templates by doing that as you have to use `${{ secrets }}` instead. But maybe that's OK @Rugvip thoughts?

**Comment by benjdlambert** [comment]
I've put together this: https://github.com/backstage/backstage/pull/24834 happy for some feedback to push it the other way if we think that's a better idea.

**Comment by jomeier** [comment]
> @jomeier no, with `ui:widget` passwords are unfortunately not treated as secure, they are not redacted in logs, and could be accessed / logged through API's. ~When you use `ui:field: Secret` you will need to use `${{ secrets.PARAM_NAME }}` not `${{ parameters.PARAM_NAME }}` instead.~
> 
> The stars that you see in the ui input gives the sense that these things are secure and they are not, it's just a mask, so we ant to make that more visible to people that using `ui:widget: Password` things aren't secure.
> 
> EDIT: sorry I realised you were talking about form context, I think that we'd rather encourage the use of `ui:field: Secret` and then you can use the hook `useTemplateSecrets` to grab the secrets context rather than using the `formContext` for this.

Can I use `TemplateSecrets` in a seperate custom UI field to access secrets from another ui field?

**Comment by jomeier** [comment]
> I'm gonna whip up a PR to replace the `ui:widget: password` with our our own `secret` implementation, but I have some concerns that we're going to break existing templates by doing that as you have to use `${{ secrets }}` instead. But maybe that's OK @Rugvip thoughts?

I'm sure it would break templates :D

**Comment by benjdlambert** [comment]
> Can I use TemplateSecrets in a seperate custom UI field to access secrets from another ui field?

Yes, for sure!

> I'm sure it would break templates :D

Yeah, but is that better than end users possibly using `ui:widget: password` and potentially leaking sensitive data through API's?

**Comment by acierto** [comment]
With a proper guidelines how to fix it, it shouldn't be a big burden how to update templates. 
Especially if `upgrade` command will point which templates are affected

**Comment by benjdlambert** [comment]
Think we can close this at this point. Should be covered by the `ui:field: Secret` instead.
